src/platforms/common/swdptap.c

```diff
@@ -132,43 +160,66 @@ static bool swdptap_seq_in_parity(uint32_t *ret, size_t clock_cycles)
 	for (volatile uint32_t counter = target_clk_divider + 1; counter > 0; --counter)
 		continue;
 
-	const bool parity = calculate_odd_parity(result);
 	const bool bit = gpio_get(SWDIO_IN_PORT, SWDIO_IN_PIN);
 
 	gpio_set(SWCLK_PORT, SWCLK_PIN);
 	for (volatile uint32_t counter = target_clk_divider + 1; counter > 0; --counter)
 		continue;
 
-	*ret = result;
+	gpio_clear(SWCLK_PORT, SWCLK_PIN);
 	/* Terminate the read cycle now */
 	swdptap_turnaround(SWDIO_STATUS_DRIVE);
+
+	const bool parity = calculate_odd_parity(result);
+	*ret = result;
 	return parity != bit;
 }
 
 static void swdptap_seq_out_clk_delay(uint32_t tms_states, size_t clock_cycles) __attribute__((optimize(3)));
 
 static void swdptap_seq_out_clk_delay(const uint32_t tms_states, const size_t clock_cycles)
 {
-	for (size_t cycle = 0; cycle < clock_cycles; ++cycle) {
-		gpio_clear(SWCLK_PORT, SWCLK_PIN);
-		gpio_set_val(SWDIO_PORT, SWDIO_PIN, tms_states & (1U << cycle));
+	uint32_t value = tms_states;
+	bool bit = value & 1;
```

Comment:
Almost all the places bit masking is being done are fine, but this and the 3 other `& 1`'s below are missing their `U` suffix to make sure this stays unsigned and doesn't cast. This should be a signed-unsigned conversion warning were `-Wconversion`/`-Warith-conversion` to be enabled.

Reply:
The left operand is a `uint32_t`. My understanding is that the only way the C arithmetic conversion rules work, the only way that literal un-suffixed integer `1` can convert the left-hand operand to a signed type is if `uint32_t` can't represent all the values of `int`, which is unlikely on a 32-bit platform.

I can add the `U` suffix anyway if that helps suppress warnings or makes it easier to manually analyze.

Reply:
The way the rules work, mean that if either side doesn't match in signed-ness, both are converted to `int` for the operation - so `& 1` turns `value` -> `int`, and then because the operation clears the sign bit, if set, this invokes UB. All bitwise manipulation of values is best done unsigned to avoid this and all signed-unsigned conversions. The result of `value & 1` is therefore `int` while `value & 1U` is `unsigned`.

Please add the suffix to prevent this error and UB invocation which is what those warnings pick up on and warn about.

Reply:
From C2x N3054

> 6.3.1.8 Usual arithmetic conversions

[...]

> Otherwise, if the operand that has unsigned integer type has rank greater or equal to the rank of the type of the other operand, then the operand with signed integer type is converted to the type of the operand with unsigned integer type.


> Otherwise, if the type of the operand with signed integer type can represent all the values of the type of the operand with unsigned integer type, then the operand with unsigned integer type is converted to the type of the operand with signed integer type.

So the risk of unsigned-to-signed conversion is if `uint32_t` is a lower rank than `int`. I'll add the `U` suffix for future-proofing, though.

Reply:
In C11 (N1548), the relevant chunk from that starts out with:

> Otherwise, the integer promotions are performed on both operands. Then the
following rules are applied to the promoted operands:

Meaning that before any of those other "Otherwise" clauses occur, integer promotion takes place. This is governed by §6.3.1.1 par 2. Suffixing inhibits all of this completely though.

Reply:
Ah, C1x lacks the language about bit-precise types in 6.3.1.1. Anyway, integer promotions happen separately for each relevant operand. I think `uint32_t` won't promote to (signed) `int` unless `int` can represent all the values of `uint32_t`, which it won't on most 32-bit platforms. In the unlikely case that `uint32_t` does promote to (signed) `int`, adding the `U` suffix to the literal will keep both operands unsigned during the usual arithmetic conversions.

Reply:
Correct, bit-precise was added in 23; and well, that promotion occurring has repeatedly bitten BMD, which is why the campaign to U-suffix all bitwise manipulation as it's fixed all kinds of bugs and miscompilations/misoptimisation allowed by the resulting UB invoked when accidentally doing bad things to the sign bit, accidentally doing `asr` instead of `lsr`, and so on. Suffixing makes the types all agree and inhibits that class of problem for cheap. We'd prefer to over-apply it than under, even if the standard says in this narrow case that it's Probably Ok™.